Added a How-To guide for qasm2 (#10557)
Qiskit/qiskit-terra: docs/how_to/use_qasm2.rst

################################
Import and export using QASM 2.0
################################

This guide shows you how to import and export quantum circuits using ``QASM 2.0``.

QASM 2.0 is short for Quantum Assembly Language version 2.0. It's a programming language specifically
designed for representing universal quantum computing using the circuit model.

You can find more information in the  `OpenQASM 2.0 repository <https://github.com/openqasm/openqasm/tree/OpenQASM2.x>`_.

Design a quantum circuit
========================

In this section, you'll build a four-qubit quantum circuit using only CX (CNOT), H (Hadamard), and RZ gates.
This circuit includes a custom quantum gate, created using the :meth:`~qiskit.QuantumCircuit.to_gate` method.
The :meth:`~qiskit.QuantumCircuit.to_gate` method allows us to convert a part of the quantum circuit into a single gate,
which acts like a building block encapsulating the circuit's functionality.

.. plot::
    :include-source:

    from qiskit import QuantumCircuit
    import numpy as np

>     custom_gate = QuantumCircuit(1) # Adds a custom gate to qubit 1
>     custom_gate.h(0)
>     custom_gate.rz(np.pi,0)
>     custom_gate.h(0)
> 
>     gate = custom_gate.to_gate() # Converts the custom_gate() into a gate.
>     gate.name = 'custom X'
> 
>     qc = QuantumCircuit(4,4)
>     qc.h(0)
>     qc.cx(0,1)
>     qc.cx(1,2)
>     qc.cx(2,3)
>     qc.append(gate,[1])  # To add a custom gate, you need to use the method append.

Reviewer: ```suggestion
    custom_gate = QuantumCircuit(1)
    custom_gate.h(0)
    custom_gate.rz(np.pi,0)
    custom_gate.h(0)

    gate = custom_gate.to_gate() # Converts the custom_gate() into a gate.
    gate.name = 'custom X'

    qc = QuantumCircuit(4,4)
    qc.h(0)
    qc.cx(0,1)
    qc.cx(1,2)
    qc.cx(2,3)
    qc.append(gate,[1])   # Adds a custom gate to qubit 1
```

>     qc.append(gate,[1])  # To add a custom gate, you need to use the method append, where you indicate the gate and the QuantumReigsters.

Reviewer: The comments are a bit too long, can you make them more concise? Or split them over a couple of lines?

![Screenshot 2023-08-11 at 14 21 33](https://github.com/Qiskit/qiskit-terra/assets/36071638/a6236704-bd88-41cd-8b09-3a08d358aed3)

For example:

```suggestion
    qc.append(gate,[1])  # Adds a custom gate to qubit 1
```

.. testcode::

    qc.qasm()

.. testoutput::

    'OPENQASM 2.0;\ninclude "qelib1.inc";\ngate custom_X q0 { h q0; rz(pi) q0; h q0; }\nqreg q[4];\ncreg c[4];\nh q[0];\ncx q[0],q[1];\ncx q[1],q[2];\ncx q[2],q[3];\ncustom_X q[1];\nbarrier q[0],q[1],q[2],q[3];\nmeasure q[0] -> c[0];\nmeasure q[1] -> c[1];\nmeasure q[2] -> c[2];\nmeasure q[3] -> c[3];\n'

You can print a formatted string by setting the Boolean `formatted` parameter to `True` (default value is `False`).

.. testcode::

    qc.qasm()

.. testoutput::

    OPENQASM 2.0;
    include "qelib1.inc";
    gate custom_X q0 { h q0; rz(pi) q0; h q0; }
    qreg q[4];
    creg c[4];
    h q[0];
    cx q[0],q[1];
    cx q[1],q[2];
    cx q[2],q[3];
    custom_X q[1];
    barrier q[0],q[1],q[2],q[3];
    measure q[0] -> c[0];
    measure q[1] -> c[1];
    measure q[2] -> c[2];
    measure q[3] -> c[3];

You can even save the code in a QASM file by using the ``filename`` parameter.
This parameter should be a string where you specify the name and follow it with the ``.qasm`` extension.

.. testcode::

    qc.qasm(filename='example.qasm')

.. testoutput::

>     'OPENQASM 2.0;\ninclude "qelib1.inc";\ngate custom_X q0 { h q0; rz(pi) q0; h q0; }\nqreg q[4];\ncreg c[4];\nh q[0];\ncx q[0],q[1];\ncx q[1],q[2];\ncx q[2],q[3];\ncustom_X q[1];\nbarrier q[0],q[1],q[2],q[3];\nmeasure q[0] -> c[0];\nmeasure q[1] -> c[1];\nmeasure q[2] -> c[2];\nmeasure q[3] -> c[3];\n'

Reviewer: Does `qc.qasm(filename='example.qasm')` produce output?

Reviewer: Just checked; turns out it does 🤷

################################
Import and export using QASM 2.0
################################

This guide shows you how to import and export quantum circuits using the ``QASM 2.0`` program.

QASM 2.0 is short for Quantum Assembly Language version 2.0. It's a programming language specifically
designed for representing universal quantum computing using the circuit model.

You can find more in its repository `OpenQASM 2.0 <https://github.com/openqasm/openqasm/tree/OpenQASM2.x>`_.

Design a quantum circuit
==============================

> Build a 4 qubits quantum circuit using only CX (CNOT), H (Hadamard), and RZ gates.

Reviewer: ```suggestion
In this section, you'll build a four-qubit quantum circuit using only CX (CNOT), H (Hadamard), and RZ gates.
```

> This circuit works with a custom quantum circuit, this is possible with the :meth:`~qiskit.QuantumCircuit.to_gate` method.

Reviewer: ```suggestion
This circuit uses a custom gate, created using the :meth:`~qiskit.QuantumCircuit.to_gate` method.
```

It is possible to format the string in an OpenQASM format using the boolean parameter `formatted`,
with its default value being False. If you set it to True, you will be able to see the changes.

.. testcode::

    qc.qasm()

.. testoutput::

    OPENQASM 2.0;
    include "qelib1.inc";
    gate custom_X q0 { h q0; rz(pi) q0; h q0; }
    qreg q[4];
    creg c[4];
    h q[0];
    cx q[0],q[1];
    cx q[1],q[2];
    cx q[2],q[3];
    custom_X q[1];
    barrier q[0],q[1],q[2],q[3];
    measure q[0] -> c[0];
    measure q[1] -> c[1];
    measure q[2] -> c[2];
    measure q[3] -> c[3];

You can even save the code in a QASM file by using the ``filename`` parameter.
This parameter should be a string where you specify the name and follow it with the ``.qasm`` extension.

.. testcode::

    qc.qasm(filename='example.qasm')

.. testoutput::

    'OPENQASM 2.0;\ninclude "qelib1.inc";\ngate custom_X q0 { h q0; rz(pi) q0; h q0; }\nqreg q[4];\ncreg c[4];\nh q[0];\ncx q[0],q[1];\ncx q[1],q[2];\ncx q[2],q[3];\ncustom_X q[1];\nbarrier q[0],q[1],q[2],q[3];\nmeasure q[0] -> c[0];\nmeasure q[1] -> c[1];\nmeasure q[2] -> c[2];\nmeasure q[3] -> c[3];\n'

> Import OpenQASM 2 to QuantumCircuit
> ====================================

Reviewer: ```suggestion
Import OpenQASM 2 to a quantum circuit
======================================
```

Qiskit has a specific module called qiskit-qasm2 that helps to import OpenQASM 2.0 files.

.. note::
    You can install the module using the following command

    ``pip install qiskit-qasm2``

> There are two methods available for parsing OpenQASM 2 programs. One method can be used to read a string, and the other method can be used to read a qasm file.

Reviewer: ```suggestion
There are two methods available for parsing OpenQASM 2 programs. One method can be used to read a string, and the other method can be used to read a QASM file.
```
To keep consistent with line 85
